cmd/api/discovery/apiresources_test.go

```diff
@@ -33,7 +42,33 @@ func TestGetAPIResource(t *testing.T) {
 
 	k8sClient := fakeK8s.NewSimpleClientset()
 	k8sClient.Resources = testAPIResourceList
-	fakeDiscovery := k8sClient.Discovery()
+	fakeDiscovery := &fakeRest.RESTClient{
+		NegotiatedSerializer: resource.UnstructuredPlusDefaultContentConfig().NegotiatedSerializer,
+		Client: fakeRest.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
+			switch req.Method {
+			case http.MethodGet:
+				switch req.URL.Path {
+				case "/apis/invalid/v1": // Invalid group
+					return &http.Response{StatusCode: http.StatusNotFound}, nil
+				case "/apis/stable.example.com/v2": // Invalid version
+					return &http.Response{StatusCode: http.StatusNotFound}, nil
+				case "/apis/stable.example.com/v1": // Valid resource
+					bodyBytes, err := json.Marshal(testAPIResourceList[0])
+					if err != nil {
+						t.Fatalf(err.Error())
+					}
+					body := io.NopCloser(bytes.NewReader(bodyBytes))
+					return &http.Response{StatusCode: http.StatusOK, Body: body}, nil
+				default:
+					t.Fatalf("unexpected request: %#v\n%#v", req.URL, req)
+					return nil, nil
+				}
+			default:
+				t.Fatalf("unexpected request: %s %#v\n%#v", req.Method, req.URL, req)
+				return nil, nil
+			}
+		}),
+	}
```

Comment:
I also played around with how to mock the restClient and I came out with a different solution.
I wanted to take advantage of the faked `ServerResourcesForGroupVersion` that was being used before, so I call it and mock the RestClient after that:

```suggestion
           // call ServerResourcesForGroupVersion and properly set the response and the status of the server
			apiResources, err := fakeDiscovery.ServerResourcesForGroupVersion(fmt.Sprintf("%s/%s", tc.group, tc.version))
			status := http.StatusOK
			response := ""
			if err != nil {
				response = err.Error()
				status = http.StatusBadRequest
			} else {
				buf := bytes.NewBufferString("")
				serializer.Encode(apiResources, buf)
				response = buf.String()
			}
           // mock the restClient with the given status and response
			restClient := &fakerest.RESTClient{
				Client: fakerest.CreateHTTPClient(func(request *http.Request) (*http.Response, error) {
					resp := &http.Response{
						StatusCode: status,
						Body: io.NopCloser(
							strings.NewReader(response)),
					}
					return resp, nil
				}),
				NegotiatedSerializer: scheme.Codecs.WithoutConversion(),
				GroupVersion:         schema.GroupVersion{Version: "v1"},
				VersionedAPIPath:     fmt.Sprintf("/apis/%s/%s", tc.group, tc.version),
			}
```

For that to work is needed also the serializer:

```
	s := runtime.NewScheme()
	serializer := json.NewSerializerWithOptions(json.DefaultMetaFactory, s, s, json.SerializerOptions{
		Pretty: true,
		Yaml:   false,
		Strict: true,
	})
```

What I like of this solution is that the logic is mostly deferred to the fake module, I just put in place the glue between the faked `ServerResourcesForGroupVersion` and the RestClient

Reply:
Good idea! I went with this idea but dropped the serializer in favor of just using `json.Marshal`.

Reply:
Nice! I like how it looks now :)